File: src/main/java/nz/co/aetheric/sameness/equatators/NullsafeEquatator.java
package nz.co.aetheric.sameness.equatators;

import javax.validation.constraints.NotNull;

/**
 * TODO: What is the purpose of this class?
 * <p>Author: <a href="http://gplus.to/tzrlk">Peter Cummuskey</a></p>
 */
@SuppressWarnings("WeakerAccess")
public class NullsafeEquatator<T> extends TypedEquatator<T> {

	public static <T> NullsafeEquatator<T> nullsafeEquatator(@NotNull Class<T> type) {
		return new NullsafeEquatator<>(type);
	}

	public NullsafeEquatator(@NotNull Class<T> type) {
		super(type);
	}

	@Override
	public <L extends T, R extends T> boolean isEquatable(L left, R right) {
		return bothNull(left, right)
				|| neitherNull(left, right);
	}

	public <L extends T, R extends T> boolean bothNull(L left, R right) {
		return left == null && right == null;
	}

	public <L extends T, R extends T> boolean neitherNull(L left, R right) {
		return left != null && right != null;
	}

}

File: src/main/kotlin/nz/co/aetheric/sameness/equatators/NullsafeEquatator.kt
package nz.co.aetheric.sameness.equatators

/**
 * TODO: What is the purpose of this class?
 *
 * Author: [Peter Cummuskey](http://gplus.to/tzrlk)
 */
open class NullsafeEquatator<T>(type: Class<T>) : TypedEquatator<T>(type) {

	companion object {

		fun <T> nullsafeEquatator(type: Class<T>)
				= NullsafeEquatator(type)

	}

	final override fun <L : T, R : T> isEquatable(left: L?, right: R?)
			= ( left == null && right == null )
			|| ( left != null && right != null )
			&& isSafelyEquatable(left, right)

	open fun <L:T,R:T> isSafelyEquatable(left: L, right: R) = true

}
